### ISSUE TITLE
Overlapping views during Transition

### ISSUE THREAD

**Issue by mikehwagz** [Dogstudio/highway#17]
Hey! 👋 great work on this library! Love the motivation and overall direction.

I noticed something while reading over the source code this evening though that would prevent me from using highway:

Currently, it seems that the `add` and `remove` methods of `Highway.Renderer` overwrite the `innerHTML` of the view element. This would make it impossible to create a transition where the old view and the new view are both visible on screen at the same time (i.e. the new view sliding in over the old view).

ideally, the new view would be appended to the wrapper, and the old view would not be removed until it is fully hidden.

i want to adopt highway, but unfortunately, the absence of this feature is a deal breaker for me. i would love to contribute and help with adding this feature, so let me know what you think!

cheers 🍻

**Comment by Anthodpnt** [comment]
Hey @mikehwagz !

You're right, good catch, we are trying to think about all possible scenarios but we are missing some indeed and that's why contributions like yours is important to us.

We are going to work on a solution for that as soon as possible and keep you in touch when it'll be released. We are going to add the `data-router-view` earlier in the process, when the old one starts to hide and you'll have to style it the way you need to display properly using `position: absolute` or something else to fit the needs of your transition... Does it look good to you ?

Best regards,
Anthodpnt

**Comment by Anthodpnt** [comment]
The problem I see though is that to do this kind of transition we have to fetch the next page before we start to hide the new one. This means that on slow networks when users will click a link the fetching could take a bit of time leaving the user with no response to the action he did and your site will just look buggy...

**Comment by mikehwagz** [comment]
Hey @Anthodpnt! Thanks so much for the quick responses.

Because this feature is only necessary for certain kinds of transitions, perhaps it could be an opt-in feature via an option.

At the same time, I understand that a primary goal of Highway is simplicity and adding config options could easily become a mess over time.

**Comment by Anthodpnt** [comment]
Yes that's right, but making an opt-in option for this that change the all process of the transition could result in pretty ugly and hard to maintain scripts. Another goal is to keep it lightweight so if we have to add lot of conditional statements to take this option into account we make miss this goal to.

I really understand your need here but after thinking about it doesn't seem to be a good idea yet. We would be really sad if you don't give a try to Highway because of one transition while there is a bunch of possibilities.

However as you can see I labelled this issue with `help wanted` and I'll leave it open so that maybe someone out there have THE solution to meet our goal and give you this feature you as. Until then we are going to keep thinking about that and keep you in touch if we have a solution.

Thanks a lot for your feedback,
Anthodpnt

**Comment by mikehwagz** [comment]
Completely understood! I respect the desire to keep this as simple and lightweight as possible.

Thanks for all your feedback. I'll definitely be thinking about it as well.

**Comment by michaelkoelewijn** [comment]
@Anthodpnt @mikehwagz any update on this? This is a feature i require before i make the decision to incorporate this plugin to my projects. It looks amazing already, but a lot of my transitions need both views to be active at the same time.

**Comment by ThaoD5** [comment]
@michaelkoelewijn glad to hear you consider using Highway for your project,

Unfortunately, for this "request", it is not near-to-be implemented yet, as it didn't really progress about this since @Anthodpnt 's last message.

As said, we're looking for solutions on this & inputs for a potential way of allowing this feature without having to completely modify the lightweight-core of Highway / add a second "core".

So if you need this feature really soon in order to use on a project that is going live soon, there may be a lot of chances that we're still figuring it out, or working on it.

I hope this will not prevent you from using Highway on other projects :-) 

Have a great day,
Thao

**Comment by michaelkoelewijn** [comment]
@ThaoD5 Thanks for your reply.
I haven't used Highway in any of my projects yet as i'm still looking for that "best" plugin to use.
Currently i'm using Barba.js a lot but Highway looks like it does the job faster and the code is easier (kudos for that).

I hope this feature can be added to the project. If not, i'll still come back if my art director decides to think of a page animation that relies on an overlay (like in your demos). Good luck with the project!

**Comment by joshkirk-zero** [comment]
Hi folks, I've added this, as well as the ability to fire a custom/contextual transition on click. Grab the built files here: https://github.com/joshkirk-zero/route66. See https://github.com/Dogstudio/highway/pull/34 for an explanation on firing contextual transitions on click and https://github.com/Dogstudio/highway/pull/39 for the change that allows both the old and new view to overlap long enough to achieve the visual effects you want. Much thanks and 99% of the credit goes to @Anthodpnt.

**Comment by joshkirk-zero** [comment]
Note that if you use the route66 version and intend to have persistent elements across views (as opposed to having a loader mask cover the views), you should place the `window.scrollTo(0, 0);` after the `done` in the `in(view, done) {` of your transition to avoid weird scroll jumps. 

Also, for the Belgians in the crowd, Route 66 is a famous highway in the USA that runs from Chicago to California: https://en.wikipedia.org/wiki/U.S._Route_66

**Comment by Anthodpnt** [comment]
Thanks a lot @joshkirk-zero !

I have merged the pull request and I'll analyze and try it on the `develop` branch.
I'll keep you in touch when it'll be released with the contextual transitions :).

**Comment by michaelkoelewijn** [comment]
I see great progress here!
What's the current status?

**Comment by joshkirk-zero** [comment]
Hey all, went back and edited my previous responses, which included stop-gap solution in the form of built highway files with slight modifications. Instead, a new release is coming that allows for everything discussed here + more! 👍

**Comment by michaelkoelewijn** [comment]
> Hey guys,
> 
> The last release [**v2.1.0**](https://github.com/Dogstudio/highway/releases/tag/v2.1.0) allows now to create overlapping transitions.
> You can get an [example](https://dogstudio.github.io/highway/examples/transitions.html#contextual-transition) right here.
> 
> Enjoy!
> Anthodpnt

Awesome! I'll give this a shot in a demo project. If all works as expected i'm using this piece of code in future projects!
